Annotate Azure.Messaging.EventHubs for trim/AOT compat

sbomer opened the pull request:
Adds annotations to resolve trim/AOT warnings in Azure.Messaging.EventHubs.

Adds this project to the AOT ci checks, and also opts the project into IsTrimmable and IsAotCompatible to get analyzer warnings.

sbomer committed:
Annotate Azure.Messaging.EventHubs for trim/AOT compat

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Messaging.EventHubs](https://apiview.dev/Assemblies/Review/050a8e78e2e94b92b38ec4920307efda?revisionId=9e5de149baac464bb21796586f40dc76)

m-redding commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Azure.Messaging.EventHubs.csproj:
```diff
@@ -7,6 +7,8 @@
     <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <IsTrimmable Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net6.0'))">true</IsTrimmable>
+    <IsAotCompatible Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0'))">true</IsAotCompatible>
```
What do these do without a net6.0/net7.0 target?

eerhardt replied:
I'm on the fence on whether we should make this change or not. It feels like either we should build for a compatible TFM or drop these properties.

The concern is that we check this in now, some one adds warnings to the assembly but that doesn't get caught since these analyzers aren't running. Then when the next person tries to add a `net8.0` TFM they are blocked because these analyzers start emitting warnings.

jsquire replied:
@eerhardt: For context, we're in the process of adding a `net8` target across the Azure SDK packages, so this will be active in the short term.

eerhardt replied:
That's good information. Given that, then I would lean towards leaving these lines in.

jsquire reviewed (commented):
Thanks very much for your contribution, @sbomer.    Would you be open to making the same changes to the `Azure.Messaging.EventHubs.Processor` package?   We like these to evolve as a unit and offer an equivalent contract.

Would you also please add a change log entry?

sbomer committed:
PR feedback

- Annotate Azure.Messaging.EventHubs.Processor, add ci check
- Add changelog entries

Also make some minor fixes to the AOT compat script.

jsquire reviewed (approved):
Thanks, @sbomer!

jsquire commented:
@m-redding:  Would you do a final pass over the new changes and merge if things look good to you?

eerhardt commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/EventHubClientBuilderExtensions.cs:
```diff
@@ -46,6 +47,8 @@ public static IAzureClientBuilder<EventHubProducerClient, EventHubProducerClient
         ///   Registers a <see cref="EventHubProducerClient"/> instance with connection options loaded from the provided <paramref name="configuration"/> instance.
         /// </summary>
         ///
+        [RequiresUnreferencedCode("Binding strongly typed objects to configuration values requires generating dynamic code at runtime, for example instantiating generic types. Use the Configuration Binder Source Generator (EnableConfigurationBindingGenerator=true) instead.")]
```
> requires generating dynamic code at runtime

(nit) This RequiresUnreferencedCode attribute has a message saying it requires generating dynamic code.

eerhardt reviewed (approved):
My 2 comments shouldn't block this PR. Overall this change looks good to me.

sbomer committed:
Fix RequiresUnreferencedCode message

m-redding commented:
This looks good to me! Thank you!! @sbomer
